File: app/src/main/java/opencrowd/hgc/hgcwallet/crypto/EDKeyChain.java
package opencrowd.hgc.hgcwallet.crypto;

import android.support.annotation.NonNull;

public class EDKeyChain implements KeyChain {
    private HGCSeed hgcSeed;

    public EDKeyChain(HGCSeed seed){
        hgcSeed = seed;
    }

    @NonNull
    @Override
    public KeyPair keyAtIndex(long index) {
        byte[] edSeed = CryptoUtils.deriveKey(hgcSeed.getEntropy(), index, 32);
        EDKeyPair pair = new EDKeyPair(edSeed);
        return pair;
    }
}

File: app/src/main/java/hedera/hgc/hgcwallet/crypto/EDKeyChain.kt


package hedera.hgc.hgcwallet.crypto

import hedera.hgc.hgcwallet.crypto.bip39.Mnemonic

class EDKeyChain(private val hgcSeed: HGCSeed) : KeyChain {

    override fun keyAtIndex(index: Int): KeyPair {
        val edSeed = CryptoUtils.deriveKey(hgcSeed.entropy, index.toLong(), 32)
        return EDKeyPair(edSeed)
    }
}

class EDBip32KeyChain(private val hgcSeed: HGCSeed) : KeyChain {
    override fun keyAtIndex(index: Int): KeyPair {
        val words = hgcSeed.toWordsList().joinToString(" ")
        val seed = Mnemonic.generateSeed(words, "")
        val ckd = SLIP10.deriveEd25519PrivateKey(seed, 44, 3030, 0, 0, index)
        return EDKeyPair(ckd)
    }
}
